airbyte-cdk/bulk/toolkits/load-dlq/src/main/kotlin/io/airbyte/cdk/load/message/dlq/DlqRecordBuilder.kt

```diff
@@ -16,18 +21,46 @@ import java.util.UUID
 /**
  * A helper method to modify a existing DestinationRecordRaw for a dead letter queue.
  *
+ * If [keepOriginalFields] is true, the data will be appended to the data blob. Otherwise, the data
+ * will replace the entire data blob of the record.
+ *
  * This method ensures we still track the correct metadata against the original record when it comes
  * to stats and checkpointing while allowing a different content to be persisted to the dead letter
  * queue.
  */
-fun DestinationRecordRaw.toDlqRecord(data: Map<String, Any>): DestinationRecordRaw =
+fun DestinationRecordRaw.toDlqRecord(
+    data: Map<String, Any>,
+    keepOriginalFields: Boolean = true,
+): DestinationRecordRaw =
     // We want to preserve everything from the original record and only override the content.
     // The rationale is that we should still be reporting original byte size and adhere to the
     // original checkpoint.
     copy(
-        rawData = DestinationRecordJsonSource(data.toAirbyteRecordMessage()),
+        rawData =
+            if (keepOriginalFields) merge(this.rawData, data)
+            else DestinationRecordJsonSource(data.toAirbyteRecordMessage()),
     )
 
+private fun merge(raw: DestinationRecordSource, data: Map<String, Any>): DestinationRecordSource {
+    when (raw) {
+        is DestinationRecordJsonSource -> {
+            val objectNode = raw.source.record.data as ObjectNode
+            data.forEach { (k, v) ->
+                objectNode.set<JsonNode>(k, Jsons.convertValue(v, JsonNode::class.java))
+            }
+        }
+        is DestinationRecordProtobufSource -> {
+            data.forEach { (_, v) ->
+                val airbyteVal = AirbyteValue.from(v)
```

Comment:
Should we use the method `asEnrichedDestinationRecordAirbyteValue` to retrieve the type here?

Reply:
isn't enriched doing all the type coersion? What I want is the record as it was, not extra type mods or anything just so that I can append extra columns.